include/cadmium/celldevs/cell/cell.hpp

```diff
@@ -0,0 +1,224 @@
+        cell(){ throw std::exception(); }
+
+        /**
+         * @brief Creates a new cell with neighbors which vicinity is set to the default.
+         * @tparam Args type of any additional parameter required for creating the output delayer.
+         * @param cell_id ID of the cell to be created.
+         * @param initial_state initial state of the cell.
+         * @param neighbors_in vector containing the ID of the neighboring cells.
+         * @param delayer_id ID of the output delayer used by the cell.
+         * @param args output delayer buffer additional initialization parameters.
+         */
+        template <typename... Args>
+        cell(C const &cell_id, S initial_state, std::vector<C> const &neighbors_in, std::string const &delayer_id, Args&&... args) {
+            NV vicinity = NV();
+            for (auto const &neighbor: neighbors_in) {
+                vicinity.insert({neighbor, V()});
+            }
+            new (this) cell(cell_id, initial_state, vicinity, delayer_id, std::forward<Args>(args)...);
+        }
+
+        /**
+         * @brief Creates a new cell with neighbors which vicinity is explicitly specified.
+         * @tparam Args type of any additional parameter required for creating the output delayer.
+         * @param cell_id ID of the cell to be created.
+         * @param initial_state initial state of the cell.
+         * @param vicinity unordered map which key is a neighboring cell and its value corresponds to the vicinity
+         * @param delayer_id ID of the output delayer buffer used by the cell.
+         * @param args output delayer buffer additional initialization parameters.
+         */
+        template <typename... Args>
+        cell(C const &cell_id_in, S initial_state, NV const &vicinity, std::string const &delayer_id, Args&&... args) {
+            assert(std::numeric_limits<T>::has_infinity && "This time base does not define infinity");
+            cell_id = cell_id_in;
+            clock = T();
+            next_internal = T();
+            state.current_state = initial_state;
+            for (auto const &entry: vicinity) {
+                neighbors.push_back(entry.first);
+                state.neighbors_vicinity[entry.first] = entry.second;
+                state.neighbors_state[entry.first] = S();
+            }
+            buffer = delayer_factory<T, S>::create_delayer(delayer_id, std::forward<Args>(args)...);
+            buffer->add_to_buffer(initial_state, T());  // At t = 0, every cell communicates its state to its neighbors
+        }
+
+        /************** USER-DEFINED METHODS **************/
+        /// @return cell's next state.
+        virtual S local_computation() const { return state.current_state; }
+        /// @return delay to be applied before communicating to neighbors a new state.
+        virtual T output_delay(S const &cell_state) const { return std::numeric_limits<T>::infinity(); }
```

Comment:
If you expect user to define them maybe its better to mark them virtual than give default implementations, else it will be hard to debug when someone falls into the default by mistake.